tracecat/auth/saml.py

```diff
@@ -0,0 +1,289 @@
+    )
+
+    # Create a temporary file
+    with tempfile.NamedTemporaryFile(
+        mode="w+", suffix=".xml", delete=False
+    ) as tmp_file:
+        # Write the XML to the temporary file
+        tree = ET.ElementTree(root)
+        tree.write(tmp_file, encoding="unicode", xml_declaration=True)
+        tmp_file.flush()
+        tmp_file_path = tmp_file.name
+
+    try:
+        yield tmp_file_path
+    finally:
+        # Clean up the temporary file
+        os.unlink(tmp_file_path)
+
+
+def create_saml_client() -> Saml2Client:
+    saml_settings = {
+        "strict": True,
+        # The global unique identifier for this service provider
+        "entityid": TRACECAT__PUBLIC_API_URL,
+        "xmlsec_binary": XMLSEC_BINARY_PATH,
+        # Service provider settings
+        "service": {
+            "sp": {
+                "name": "tracecat_saml_sp",
+                "description": "Tracecat SAML SSO Service Provider",
+                "endpoints": {
+                    "assertion_consumer_service": [
+                        (SAML_SP_ACS_URL, BINDING_HTTP_POST),
+                    ],
+                },
+                # Security settings
+                "allow_unsolicited": True,  # If true, it allows the IdP to initiate the authentication (not yet supported)
+                "authn_requests_signed": False,  # Don't need to sign authn requests because we don't control the IdP
+                "want_assertions_signed": True,  # We require the IdP to sign the assertions
+                "want_response_signed": False,
+            },
+        },
+    }
+
+    if SAML_IDP_METADATA_URL is None:
+        with generate_saml_metadata_file() as tmp_metadata_path:
+            # Add the local metadata file to the settings
+            saml_settings["metadata"] = {"local": [tmp_metadata_path]}
+            config = Saml2Config()
+            config.load(saml_settings)
+    else:
+        # Save the cert to a temporary file
+        with tempfile.NamedTemporaryFile(
+            mode="w+", suffix=".crt", delete=False
+        ) as tmp_file:
+            tmp_file.write(SAML_IDP_CERTIFICATE)
+            tmp_file.flush()
+            saml_settings["metadata"] = {
+                "remote": [
+                    {
+                        "url": SAML_IDP_METADATA_URL,
+                        "cert": tmp_file.name,  # Path to cert
+                    }
+                ]
+            }
+            config = Saml2Config()
+            config.load(saml_settings)
+
+    client = Saml2Client(config)
+    return client
+
+
+@router.get("/login", name=f"saml:{auth_backend.name}.login")
+async def login():
+    client = create_saml_client()
+    _, info = client.prepare_for_authenticate()
+    try:
+        headers = info["headers"]
+        # Select the IdP URL to send the AuthN request to
+        for k, v in headers:
+            if k == "Location":
+                redirect_url = v
```

Comment:
nit: perhaps more concise, and it'll raise `StopIteration` if no `location` is found
```suggestion
        redirect_url = next(v for k, v in headers if k == "location")
```